fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
-    very efficient.
-    ::
+class PipelineModule(nn.Module):
+    """Constructs a module on a remote device, possibly at a later time (in case the device is not
+    specified when creating PipelineModule.
+    Args:
+        module_cls (nn.Module): Class for the module to be created remotely.
+        args (Sequence): args to be passed to ``module_cls``.
+        kwargs (Dict, optional): kwargs to be passed to ``module_cls``.
+        num_inputs (int, optional): number of inputs to the forward function.
+        num_outputs: (int, optional): If the forward function returns a tuple, number of elements
+            in the tuple, otherwise it should be None
+        remote_device: (str, optional): Device on the destination worker where we‘d like to place
+            this module. The format should be "<workername>/<device>", where the device field can be
+            parsed as torch.device type. E.g., "trainer0/cpu", "trainer0", "ps0/cuda:0".
+            This parameter can be provided later by calling the method instantiate.
+    """
+
+    def __init__(
+        self,
+        module_cls: nn.Module,
+        args: Tuple,
+        kwargs: Optional[Dict] = None,
+        num_inputs: int = 1,
+        num_outputs: Optional[int] = None,
+        remote_device: str = None,
+    ):
+        super().__init__()
+        self.num_inputs = num_inputs
+        self.num_outputs = num_outputs
+        self.module_args = (module_cls, args, kwargs or {})
+        if remote_device is not None:
+            self.instantiate(remote_device)
+
+    @staticmethod
+    def _create_module(module_cls: Callable, args: Tuple, kwargs: Dict, device: str) -> nn.Module:
+        result: nn.Module = module_cls(*args, **kwargs)
+        result.to(device)
+        return result
+
+    def instantiate(self, remote_device: str) -> "PipelineModule":
+        worker, device = remote_device.split("/")
+        self.worker = worker
+        self.device = device
+        self.module_rref = rpc.remote(worker, PipelineModule._create_module, self.module_args + (device,))
+        return self
+
+    def get_module_rref(self) -> rpc.RRef:
+        return self.module_rref
+
+
+class PipelineModulesGraph(nn.Module):
+    """A collection of remote modules (of type PipelineModule) with connections showing how inputs
+    to the model or outputs of individual modules are use as inputs of subsequent modules.
+    The graph has a number of helper functions that add new modules to the graph and define inputs
+    to these module.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.modules_list: List = []
+        # self.inputs specifies inputs to each module in modules_list. Each input to each module
+        # is represented by a tuple (i, j). If i>=0, then the input is the j'th output of the i'th
+        # module. If i<0, the input is the j'th input to the model.
+        self.inputs: List[Optional[List[Tuple[int, int]]]] = []
```

Comment:
My suggestion is to use a python structure that holds 1) the module index 2) output of module index in 1). I think it will be more clear instead of have tuples with indices. WDYT?

Reply:
Yes, I have a PR coming next that has this (#591). Can it wait?

Reply:
Sure! Are you planning to add a python structure for other places that use tuples of ints to connect modules?

Reply:
Replaced other tuples (with more than two inputs) with NamedTuple.